nova/console/shellinaboxproxy.py

```diff
@@ -13,66 +13,111 @@
-from nova import context
-from nova import exception
-
-STATIC_FILES_EXT = ('.js', '.css', '.html', '.ico', '.png', '.gif')
-
-
-class NovaShellInaBoxProxy(NovaProxyRequestHandlerBase):
-    """Class that injects token validation routine into proxy logic."""
-
-    def __init__(self):
-        self._compute_rpcapi = None
-
-    @property
-    def compute_rpcapi(self):
-        # This is copied from NovaProxyRequestHandler, just to avoid
-        # extending that class (because it inherits
-        # websockify.ProxyRequestHandler in addition and we don't need that).
-        # For upgrades we should have a look again if anything changed there,
-        # that we might need to also include here.
-        if not self._compute_rpcapi:
-            self._compute_rpcapi = compute_rpcapi.ComputeAPI()
-        return self._compute_rpcapi
-
-    def path_includes_static_files(self):
-        """Returns True if requested path includes static files."""
-        for extension in STATIC_FILES_EXT:
-            if extension in self.path:
+
+import hashlib
+
+import configparser
+from mitmproxy.models import http
+from mitmproxy.script import concurrent
+from netlib import http as netlib_http
+from netlib import version
+from sqlalchemy import create_engine
+from sqlalchemy.sql import text
+
+STATIC_FILES_EXT = (".js", ".css", ".html", ".ico", ".png", ".gif")
+
+
+def _load_dbs():
+    config = configparser.ConfigParser()
+    config.read("/etc/nova/nova.conf")
+    api_db_url = config["api_database"]["connection"]
+    api_db_engine = create_engine(api_db_url)
+    with api_db_engine.connect() as api_db:
+        for (db_url,) in api_db.execute(
+            text("SELECT database_connection FROM cell_mappings WHERE id > 1")
+        ):
+            yield create_engine(db_url)
+
+
+_DBS = list(_load_dbs())
+
+
+@concurrent
+def request(flow):
+    if flow.error or not flow.live:
+        return
+    request = flow.request
+
+    # We only have to validate the "initial" GET
+    # on the console path, as that one returns a session-id
+    # which is validated by the backend
+
+    if request.method == "POST":  # POST gets validated by the backend
+        return
+
+    path = request.path
+    if _path_includes_static_files(path):  # Not secret
+        return
+
+    if path in [None, "", "/"]:  # The liveness probe
+        _root_response(flow)
+        return
+
+    token = flow.request.query.pop("token", None)
+
+    if not token:
+        _access_denied(flow, b"No token provided.")
+        return
+
+    token_hash = _get_sha256_str(token)
+
+    for db in _DBS:
+        if _validate_token(db, token_hash):
+            return
+
+    _access_denied(flow, b"The token has expired or is invalid.")
+
+
+def _path_includes_static_files(path):
+    """Returns True if requested path includes static files."""
+    for extension in STATIC_FILES_EXT:
+        if path.endswith(extension):
+            return True
```

Comment:
`endswith` can handle a tuple, so this could just be `return path.endswith(STATIC_FILES_EXT)`
> suffix can also be a tuple of suffixes to look for. 
https://docs.python.org/2/library/stdtypes.html?highlight=endswith#str.endswith

Reply:
Good to know. I've changed it accordingly